### PR TITLE
feat: convert office documents to HTML via pandoc

### PR DESCRIPTION
**Pull request by Kikobeats**
## What

Adds office-document conversion to `html-get`, mirroring the existing PDF/mutool branch. When a fetched document is a Pandoc-native office format, it's converted to HTML at fetch time, so the rest of the pipeline (metadata, markdown, screenshot) treats it like a regular web page.

Supported: **docx, xlsx, pptx, odt, rtf, epub**.

Format is resolved from the response content-type, falling back to the URL extension (many CDNs serve office files as `application/octet-stream`).

## Out of scope

Legacy OLE binaries (`.doc/.xls/.ppt`) and ODF spreadsheet/presentation (`.ods/.odp`) are excluded: Pandoc can't read them without a headless LibreOffice.

## Details

- `src/office.js` — content-type/extension → pandoc `--from` resolution
- `src/index.js` — office branch in `fetch()`; `pandoc` runner is injectable and disablable (`pandoc: false`), same as `mutool`; office URLs route to fetch mode
- 14 tests (unit detection + integration fetch→convert), deterministic and offline, with 6 minimal binary fixtures (68 KB)

Prod base image already ships pandoc 3.10 and mupdf-tools, so no infra change. The consuming `api` needs no code change — it runs html→markdown on the converted HTML as-is.

> Note: this branch also carries the pre-existing local `515fd00` prerender fix.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> New fetch-time subprocess and temp-file handling for untrusted URLs expands attack surface similarly to mutool; format mis-detection could route wrong bytes to Pandoc instead of PDF handling, though detection rules and tests target that edge case.
> 
> **Overview**
> **Office documents** are now converted to HTML during fetch, parallel to the existing PDF/`mutool` path, so downstream steps can treat them like normal pages. Supported formats are **docx, xlsx, pptx, odt, rtf, and epub**; legacy OLE/ODS/ODP are out of scope.
> 
> A new **`pandoc` option** (auto-detected via `which pandoc`, disable with `pandoc: false`) runs Pandoc on a temp file. **`src/office.js`** maps content-type and URL extension to the Pandoc `--from` format, with rules so real HTML/media types win over vanity `.docx` URLs, mislabeled binaries can still match by extension, and redirect URLs are considered (`res.url` then the original URL). Recognized office files **skip `mutool`** even when the response is labeled `application/pdf`. Office URLs are routed to **fetch** mode like PDFs.
> 
> **CI** installs Pandoc in main and PR workflows; **README** documents the feature.
> 
> Also bundled: **shadow-DOM prerender retry** no longer runs when **`prerender: false`** is set explicitly (new test).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c5a94f177c1dfefe3a240a47dfb2fc80bff43873. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -76,6 +78,13 @@ const fetch = PCancelable.fn(
           }
         }
 
+        const officeFormat = pandoc && getOfficeFormat({ contentType, url })
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply src/index.js]: Fixed in 780e32f. `getOfficeFormat` now resolves the extension fallback from `res.url` (final URL after redirect) instead of the original request URL. Added a regression test: 302 to a `.docx` served as `application/octet-stream` now converts correctly.

File: src/index.js [hunk 2]
```diff
@@ -219,7 +228,14 @@ const isFetchMode = url => {
...
 const defaultGetMode = (url, { prerender }) => {
-  if (prerender === false || isMediaUrl(url) || isPdfUrl(url)) return 'fetch'
+  if (
+    prerender === false ||
+    isMediaUrl(url) ||
+    isPdfUrl(url) ||
+    isOfficeUrl(url)
+  ) {
+    return 'fetch'
+  }
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply src/index.js:242]: Intentional, and consistent with the existing `isPdfUrl` behavior. `defaultGetMode` runs before the fetch, so the content-type is not yet known and only the URL extension can route an extensionless URL to fetch mode. The content-type detection in the fetch branch is the safety net for requests already in fetch mode (autoDomains, explicit `prerender: false`, or extension-present-but-`octet-stream`, which is the common CDN case). The narrow gap (no extension AND `prerender: auto`) is the same one PDF has had; closing it properly means making mode selection content-type-aware (HEAD probe or post-fetch reroute), out of scope for this PR.

File: src/office.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+const getOfficeFormatFromUrl = url => byExtension.get(getUrlExtension(url))
+
+// Resolve the Pandoc `--from` format for a document, preferring its declared
+// content-type and falling back to the URL extension (many CDNs serve office
+// files as application/octet-stream).
+const getOfficeFormat = ({ contentType, url } = {}) =>
+  byContentType.get(contentType) ||
+  (url ? getOfficeFormatFromUrl(url) : undefined)
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply src/office.js]: Fixed in c5a94f1. `getOfficeFormat` now treats an explicit web/media content-type (`text/html`, `application/xhtml+xml`, `application/json`, `image/*`, `audio/*`, `video/*`) as authoritative and returns `undefined`, so a page served at a `.docx` vanity URL is no longer misrouted to pandoc. The URL extension only decides when the content-type is non-committal (octet-stream, zip, missing). Test: `an explicit web content-type is not overridden by the url extension`.

File: src/index.js [hunk 4]
```diff
@@ -76,6 +78,14 @@ const fetch = PCancelable.fn(
           }
         }
 
+        const officeFormat =
+          pandoc && getOfficeFormat({ contentType, url: res.url })
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply src/index.js]: Fixed in c5a94f1. Extension fallback now tries a list of candidate URLs `[res.url, url]` in order, so an office extension present only on the original request URL survives a redirect to an extensionless `octet-stream` target. Test: `extension fallback survives redirect to an extensionless target`.

File: src/index.js [hunk 5]
```diff
@@ -76,6 +78,14 @@ const fetch = PCancelable.fn(
...
+        if (officeFormat) {
+          const file = getTemporalFile(res.url, officeFormat)
+          await writeFile(file.path, res.body)
+          return pandoc(officeFormat, file.path)
+        }
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply src/index.js:90]: Fixed in c5a94f1. Office detection now runs before the PDF branch, and the PDF branch is guarded with `!officeFormat`, so a recognized office file (by content-type or extension) is never fed to mutool. A `.docx` mislabeled as `application/pdf` is converted by pandoc instead of failing on mutool. Test: `office file mislabeled as application/pdf is converted by pandoc`.
